src/Mvc/Mvc.Core/src/Formatters/TranscodingReadStream.cs

```diff
@@ -0,0 +1,225 @@
+                ArrayPool<byte>.Shared.Rent(MaxByteCountForUTF8Char),
+                0,
+                count: 0);
+
+            _encoder = Encoding.UTF8.GetEncoder();
+            _decoder = sourceEncoding.GetDecoder();
+        }
+
+        public override bool CanRead => true;
+        public override bool CanSeek => false;
+        public override bool CanWrite => false;
+        public override long Length => throw new NotSupportedException();
+        public override long Position { get; set; }
+
+        public override void Flush()
+            => throw new NotSupportedException();
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            ThrowArgumentOutOfRangeException(buffer, offset, count);
+
+            var readBuffer = new ArraySegment<byte>(buffer, offset, count);
+
+            if (_overflowBuffer.Count > 0)
+            {
+                var bytesToCopy = Math.Min(count, _overflowBuffer.Count);
+                _overflowBuffer.Slice(0, bytesToCopy).CopyTo(readBuffer);
+
+                _overflowBuffer = _overflowBuffer.Slice(bytesToCopy);
+
+                // If we have any overflow bytes, avoid complicating the remainder of the code, by returning as
+                // soon as we copy any content.
+                return bytesToCopy;
+            }
+
+            var totalBytes = 0;
+            bool encoderCompleted;
+            int bytesEncoded;
+
+            do
+            {
+                // If we had left-over bytes from a previous read, move it to the start of the buffer and read content in to
+                // the segment that follows.
+                var eof = false;
+                if (_charBuffer.Count == 0)
+                {
+                    // Only read more content from the input stream if we have exhausted all the buffered chars.
+                    eof = await ReadInputChars(cancellationToken);
+                }
+
+                // We need to flush on the last write. This is true when we exhaust the input Stream and any buffered content.
+                var allContentRead = eof && _charBuffer.Count == 0 && _byteBuffer.Count == 0;
+
+                if (_charBuffer.Count > 0 && readBuffer.Count < MaxByteCountForUTF8Char && readBuffer.Count < Encoding.UTF8.GetByteCount(_charBuffer.AsSpan(0, 1)))
+                {
+                    // It's possible that the passed in buffer is smaller than the size required to encode a single
+                    // char. For instance, the JsonSerializer may pass in a buffer of size 1 or 2 which
+                    // is insufficient if the character requires more than 2 bytes to represent. In this case, read
+                    // content in to an overflow buffer and fill up the passed in buffer.
+                    _encoder.Convert(
```

Comment:
I am having difficulty understanding the need for `_overflowBuffer` here. It looks like this buffer will always only be between 1-3 bytes in length. Is that correct? Can you explain why its needed? Why is the JsonSerializer passing in a buffer of size 1 or 2?

Reply:
In general the buffer size is fairly large. However, the serializer sometimes call with a size 1 byte buffer:

![image](https://user-images.githubusercontent.com/174281/54445356-d4120180-4701-11e9-9915-f76edea14890.png)

The calling code: https://github.com/dotnet/corefx/blob/master/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs#L64-L70. I considered making an ask that would require the buffer to be at least 6 byte long, but this seemed like the right fix to make regardless. cc @stephentoub in case he's got any thoughts or suggestion here.

Reply:
> However, the serializer sometimes call with a size 1 byte buffer

This could be tweaked.  Currently the serializer is trying to fill the buffer it has before passing that along to the Utf8JsonReader, because every call to the reader has relatively expensive setup costs, and we need to read everything from the stream anyway, plus it potentially avoids an extra trip through the reader for an empty buffer when we make the final read on the stream and find that it's eof.

But there's a balance that can be struck.  This is just a matter of finding the right balance / heuristic in https://github.com/dotnet/corefx/blob/b8bc4ff80c5f7baa681e8a569d367356957ba78a/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs#L58-L62 as to how much to proactively fill the buffer.  As I noted in https://github.com/dotnet/corefx/pull/35823, "We can tweak this further in the future as needed, e.g. only making a second attempt to fill the buffer rather than doing so aggressively."